doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,914 @@
+# - GPU 1: experts_per_gpu[1]=3, holds experts 5-7 with tokens [103,45,201]
+```
+
+This requires **two independent ragged dimensions**:
+1. **Expert distribution across GPUs**: `[5, 3]` - GPU 0 has 5 experts, GPU 1 has 3 experts
+2. **Token distribution per expert**: `[127, 0, 198, 64, 89, 103, 45, 201]` - each of the 8 total experts processes different numbers of tokens
+
+Both dimensions are ragged but independent (not nested within each other). The uneven expert distribution enables load balancing, with GPU 0 handling more experts (478 tokens total) and GPU 1 handling fewer experts (349 tokens total), resulting in somewhat balanced workload.
+
+---
+
+## 4. Goals and Non-Goals
+
+### Goals
+
+- Support ragged dimensions in fusion IR for expert parallelism
+- Support multiple independent ragged dimensions per tensor for load balancing scenarios
+
+### Non-Goals
+
+- Inlining
+- Extension of single-GPU schedulers for ragged dimensions
+
+Inlining would be eventually necessary for lowering to CUDA code but should not be necessary for distributed device (multi-GPU) parallelization. Similarly, the schedulers will be left as is for now.
+
+---
+
+## 5. Requirements
+
+### Functional Requirements
+
+- **IR Representation**: Represent ragged dimensions with variable extents per batch component
+- **Offset Computation**: Compute and provide access to offsets for indexing into contiguous storage
+- **Operations Support**: Support creation and basic transformations on ragged dimensions (merge operation TBD based on expert parallelism requirements)
+- **Code Generation**: Generate correct CUDA code with offset-based indexing for ragged iteration
+- **Uniform Properties**: Enforce uniform execution properties (ParallelType, IterType) across components
+- **Integration**: Work within existing TensorView/TensorDomain infrastructure
+
+### Non-Functional Requirements
+
+- **Backward Compatibility**: No breaking changes to existing IterDomain API or behavior
+- **Maintainability**: Clear separation of ragged-specific logic from regular IterDomain code
+
+---
+
+## 6. Proposed Design
+
+### 6.1 High-Level Architecture
+
+`RaggedIterDomain` is a direct subclass of `IterDomain`, allowing it to be used anywhere an IterDomain is expected (in TensorDomain, transformations, etc.) while providing ragged-specific functionality.
+
+```
+Val
+ └── IterDomain
+      ├── (regular IterDomain instances)
+      └── RaggedIterDomain (new subclass)
+```
+
+#### The Partition Operation: Core Primitive for Nested Tensors
+
+The `partition` operation is the fundamental IterDomain-level transformation that creates ragged dimensions by splitting a regular IterDomain into non-uniform segments based on offsets.
+
+**Operation Signature:**
+```cpp
+// Static method in IterDomain class
+static std::pair<IterDomain*, RaggedIterDomain*> partition(
+    IterDomain* in,           // Input IterDomain to partition
+    TensorView* offsets       // Offset tensor defining partition boundaries
```

Comment:
We may be missing information here. Consider MoE takes a batch of sequences. Therefore,
```
dispatch input: 
   s/d
  /   \
[e, b, j1, h, d] and 

tokens_per_expert_per_device: [b, d, e]
```
How do we know which dimension holds the component extents? How do we know that the two `b`s should be mapped and the two `d`s should be mapped so we find the correct extent vector to partition?

Reply:
That's a good question. What I'm currently thinking about is to infer that from the TensorView domain. When this input TV, say `input_tv`, is partitioned, its domain should look like `[b, d, s/d]` (not sure about the order). Thus, when `partition` is called on this tensor, the call would look like `input_tv->partition(/*dimensionto_partition=*/2, /*offsets=*/[b, d, e])`. From this call, we would see that the `b` of the offset tensor would correspond to the `b` dimension of `input_tv`, etc. I *think* this would work without ambiguity, but I may be wrong, and if so, I'd make this call interface more explicit (e.g., specifying which dimension of the offset tensor correspond to which dimension of the tensor to partition).

Reply:
> Positional matching should give us enough information.

Sounds good! 

> The exact graph can be used.

We could -- it's kind of expensive because it has to analyze across TensorViews. Can we simply have RaggedIterDomain point to the IterDomain (in the same TensorView) that represents the number of components?